examples/triangle_intersection/triangle_intersection.cpp

```diff
@@ -0,0 +1,387 @@
+  {
+    ArborX::ExperimentalHyperGeometry::Point<2> const &point =
+        getGeometry(getPredicate(query));
+    auto query_index = ArborX::getData(query);
+
+    auto const coeffs =
+        _triangles.get_mapping(primitive.index).get_coeff(point);
+    bool intersects = coeffs[0] >= 0 && coeffs[1] >= 0 && coeffs[2] >= 0;
+
+    if (intersects)
+    {
+      _offsets(query_index) = primitive.index;
+      _coefficients(query_index) = coeffs;
+      return ArborX::CallbackTreeTraversalControl::early_exit;
+    }
+    return ArborX::CallbackTreeTraversalControl::normal_continuation;
+  }
+
+private:
+  Triangles<MemorySpace> _triangles;
+  Kokkos::View<int *, MemorySpace> _offsets;
+  Kokkos::View<ArborX::Point *, MemorySpace> _coefficients;
+};
+
+// Now that we have encapsulated the objects and queries to be used within the
+// Triangles class, we can continue with performing the actual search.
+int main()
+{
+  Kokkos::ScopeGuard guard;
+
+  using ExecutionSpace = Kokkos::DefaultExecutionSpace;
+  using MemorySpace = typename ExecutionSpace::memory_space;
+  ExecutionSpace execution_space;
+
+  std::cout << "Create grid with triangles.\n";
+  Triangles<MemorySpace> triangles(execution_space);
+  std::cout << "Triangles set up.\n";
+
+  std::cout << "Creating BVH tree.\n";
+  ArborX::BasicBoundingVolumeHierarchy<
+      MemorySpace, ArborX::Details::PairIndexVolume<
+                       ArborX::ExperimentalHyperGeometry::Box<2>>> const
+      tree(execution_space, triangles);
+  std::cout << "BVH tree set up.\n";
+
+  std::cout << "Create the points used for queries.\n";
+  Points<MemorySpace> points(execution_space);
+  std::cout << "Points for queries set up.\n";
+
+  std::cout << "Starting the queries.\n";
+  int const n = points.size();
+  Kokkos::View<int *, MemorySpace> offsets("Example::offsets", n);
+  Kokkos::View<ArborX::Point *, MemorySpace> coefficients(
+      "Example::coefficients", n);
+
+  tree.query(execution_space, points,
+             TriangleIntersectionCallback<MemorySpace>{triangles, offsets,
+                                                       coefficients});
+  std::cout << "Queries done.\n";
+
+#ifndef NDEBUG
+  std::cout << "Starting checking results.\n";
+  Kokkos::parallel_for(
+      Kokkos::RangePolicy<ExecutionSpace>(execution_space, 0, n),
+      KOKKOS_LAMBDA(int i) {
+        constexpr float eps = 1.e-3;
+
+        if (offsets(i) != i)
+          Kokkos::abort("Offsets are wrong");
+        auto const &c = coefficients(i);
+        auto const &t = triangles(offsets(i));
+        auto const &p_h = points(i);
+        auto const p = ArborX::ExperimentalHyperGeometry::Point<2>{
+            c[0] * t.a[0] + c[1] * t.b[0] + c[2] * t.c[0],
+            c[0] * t.a[1] + c[1] * t.b[1] + c[2] * t.c[1]};
+        if ((Kokkos::abs(p[0] - p_h[0]) > eps) ||
+            Kokkos::abs(p[1] - p_h[1]) > eps)
+          Kokkos::abort("Coefficients are wrong");
```

Comment:
Is it really right to `abort` here, especially on the first failure? I think it would make sense for the program to terminate sanely, and just print the failure.

Reply:
The program shouldn't terminate normally so we can catch regressions.

Reply:
Wouldn't non-zero return code indicate an error?

Reply:
I pushed a commit doing that. Not sure if that's better.